### ISSUE TITLE
Replace Webcontainer with more open OSS implementations

### ISSUE THREAD

**Issue by thecodacus** [coleam00/bolt.new-any-llm#229]
**Is your feature request related to a problem? Please describe:**

Currently, our project relies on Webcontainer to simulate a local development environment directly in the browser. However, Webcontainer is not open source, which limits flexibility and could pose issues for future scalability, transparency, and customization.

**Describe the solution you'd like:**

Explore fully open-source alternatives to Webcontainer that provide similar functionality while being open-source. The new solution should enable browser-based file system access, terminal simulation, real-time code execution, and preview capabilities. Additionally, it should integrate seamlessly with Vite, React, Zustand, and Dockview used in our project.

**Describe alternatives you've considered:**

1. **Wasmer**: [Wasmer](https://github.com/wasmerio/wasmer) is a blazing fast and secure WebAssembly runtime that enables incredibly lightweight containers to run anywhere: from Desktop to the Cloud, Edge and your browser.
2. **Node.js in WASM**: Running Node.js within WASM, though experimental, could replicate some Webcontainer functionalities with added flexibility.

**Additional context:**

1. Research and evaluate open-source alternatives, listing pros and cons.
2. Develop a migration plan for the chosen alternative, including testing phases to ensure compatibility with existing features.
3. Document any new configuration or setup steps required for developers.

**Comment by Mouradif** [comment]
Would love to see the preview just run locally too

**Comment by milutinke** [comment]
There are also WebVM and WasmEdge

**Comment by milutinke** [comment]
Took a look at Wasmer, apparently it doesn't support NodeJS and npm in the browser itself.
Maybe Nodebox could work: https://github.com/Sandpack/nodebox-runtime

**Comment by wonderwhy-er** [comment]
What I suggested in some conversations is not to replace outright, but rather make this a plug and play.
WebConatiners stays as default and we need to start researching replacing it with alternatives.
I am trying to reach StackBlitz in meantime to figure out weather Bolt.New with WebContaienrs requires commercial license for commercial use cases, its good to know,

I know it could be buggy, slow, challenging to pull off. 

There are multiple options,
https://webvm.io/ and https://wasmer.io/

And even more, its just that exactly for in browser development with npm and node.js WebContainers is one that was built for that, others could be slower or lack some features like npm install really working well

Potentially it could also allow using other backends rather then Node.js, some people would prefer Python for example.

**Comment by thecodacus** [comment]
I agree. I did some research on how StackBlitz likely implemented WebContainer. It turns out they didn’t put a full OS into WebAssembly. Instead, there’s no actual Wasm runtime capable of running Node.js itself. 

What they’ve done is create a detailed compatibility layer that mimics a shell environment but isn’t a true shell. This layer allows for commands like npm and other tools to run within the environment. 

They achieve this by combining WebAssembly with service workers and socket connections, effectively making the framework function seamlessly.

**Comment by wonderwhy-er** [comment]
Thanks for research @thecodacus!
Did not did deep in to how it really works as its not OSS.
Where did you dig that out?

**Comment by thecodacus** [comment]
Initially, I looked into existing efforts to compile Node.js to WebAssembly, but quickly hit a dead end. Node.js relies on certain APIs that the WASI standard doesn’t support yet. While Wasix (an extension of WASI) shows some promise—Mozilla has even compiled SpiderMonkey to Wasix—it’s still incomplete. The Wasmer team is also working on a new runtime called WinterJS for Wasix, but it doesn’t yet support all web frameworks.

StackBlitz’s official [blog post](https://blog.stackblitz.com/posts/introducing-webcontainers/) provided some additional insights. Specifically, in the sections [Run servers. In your browser.](https://blog.stackblitz.com/posts/introducing-webcontainers/#run-servers-in-your-browser) and [Zero footprint. Boots in milliseconds](https://blog.stackblitz.com/posts/introducing-webcontainers#zero-footprint-boots-in-milliseconds), they mention using a service worker for the networking layer. This is important because WebAssembly in the browser can’t make direct network calls, yet WebContainers can install packages from npm registries.

To dig deeper, I checked the available commands within WebContainer by running ls ${PATH//:/ }, but got this error: ls: cannot access '/bin:/usr/bin:/usr/local/bin': No such file or directory. This suggests that the ls command is a simplified implementation that doesn’t fully replicate standard shell functionality.

Running ls for individual paths revealed the available commands (see screenshot below). I believe many of these are lightweight implementations, with commands like npm possibly being JavaScript scripts executed in a Wasm-compatible JS runtime like QuickJS. Network requests are likely routed via service workers, and downloaded files stored in an in-memory file system.
<img width="451" alt="Screenshot 2024-11-11 at 1 12 29 AM" src="https://github.com/user-attachments/assets/cbcf32a5-d933-4539-880a-c4cb46b4d9fa">

even zsh and bash are symlink of jsh
<img width="540" alt="image" src="https://github.com/user-attachments/assets/3e3a2dd6-4107-4af0-97d3-fe4c71d37d8b">


so my final conclusion, WebContainer is a combination of simplified implementation of shell commands to operate in memory file system, and commands that would typically run in Node are instead executed in a WebAssembly-JS runtime like QuickJS, with network calls handled by service workers.

**Comment by thecodacus** [comment]
here is a section of the output of command `cat /bin/cp` from my system, i got similar result in [https://webvm.io/](https://webvm.io/) which is an actually os in browser
<img width="1272" alt="image" src="https://github.com/user-attachments/assets/d02201cf-1b07-46d3-8794-a51e4e729392">

where i got output like this for same command in webcontainer 
<img width="104" alt="image" src="https://github.com/user-attachments/assets/f4caf93e-bd9a-4e54-8410-56d7864eab3d">

and  its same for all the binaries. which also makes me believe these are just placeholders

**Comment by anbraten** [comment]
To get an idea on how webcontainers work you can cp `/bin/jsh` to `/home/project` and download it as part of the zip file. `jsh` is containing most implementation details. It contains code for the shell and also for commands like `ls` / `cat`.

**Comment by anbraten** [comment]
Did some further testing and created: https://github.com/anbraten/nano-web-ide with my current code.

**Comment by thecodacus** [comment]
I am working on this 
https://github.com/thecodacus/OpenWebContainer

**Comment by wonderwhy-er** [comment]
Woah incredible progress!

**Comment by syrusakbary** [comment]
Hey, I'm Syrus, from Wasmer.
We can get Node.js apps (and even Python) running in Wasmer. Happy to help on this effort!

**Comment by thecodacus** [comment]
Hi @syrusakbary , nice to see you here
you are talking about winterjs?

**Comment by thecodacus** [comment]
btw I was able to make the import work. only need bundle npm package without the native stuff 😄

**Comment by maitrungduc1410** [comment]
I am the author of Vivari, so treat this as disclosed self-promotion. I am
posting because this thread settled on a technical conclusion that I think is
now out of date, not because I want to advertise.

The consensus here was that running real Node in a browser is not feasible, and
that a mock or a QuickJS-style reimplementation is the only realistic path. That
was a fair reading in late 2024. There turns out to be a way around the blocker.

The blocker is that Node's API surface is synchronous while browsers refuse to
block. The exception is a Web Worker thread, where `Atomics.wait()` genuinely
parks execution. So `fs.readFileSync` parks the worker until a host answers over
a `SharedArrayBuffer`, and a kernel sitting on a Rust/Wasm virtual filesystem
services the syscall. That single primitive is what lets Node's own `lib/` run
unmodified instead of being reimplemented.

https://github.com/maitrungduc1410/vivari, MIT, no commercial license and no
usage fee.

What runs today:

- Node's real `lib/`, not a reimplementation
- npm, yarn and pnpm, installing from the real registry
- Next.js 16 App Router, server-rendering React Server Components in the tab
- Vite. SvelteKit and Nuxt are there but still experimental
- Real CPython 3.14 through Pyodide, including matplotlib, working offline
- Bun as an API-compatible shim over the Node runtime rather than the native Zig
  binary, and roughly 20 APIs refuse outright. `bun install`, `Bun.serve`,
  `Bun.build`, `Bun.plugin`, `bun:test`, `bun:sqlite` and a TypeScript REPL work
- SQLite through `bun:sqlite` and Python's `sqlite3`. Postgres through PGlite is
  still experimental
- A step debugger over the Chrome DevTools Protocol, and local-only git through
  isomorphic-git (init, stage, commit, branch, diff, history, but no clone, push
  or pull). Both are in the Studio UI rather than the SDK

Limitations, before anyone spends time on it:

- It needs `SharedArrayBuffer`, so the embedding page has to be cross-origin
  isolated (`COOP: same-origin` plus `COEP: require-corp`). That is the same
  constraint WebContainer carries, and it is the real integration cost.
- It is not a drop-in replacement for the WebContainer API. Porting an
  integration means rewriting it, not swapping an import.

If you would rather check the approach before trusting it, the two relevant
writeups are https://vivari.run/blog/nodes-real-lib-in-the-browser and
https://vivari.run/blog/blocking-in-a-browser

Live demo, no signup: https://vivari.run/studio/

Happy to answer integration questions here.
